File: reference/original-java/src/hammerwatchgen/WallPattern.java
package hammerwatchgen;

import hammerwatchgen.Doodad.DoodadType;

public class WallPattern {

    static final int w = 1; // wall
    static final int e = 2; // emtpy
    static final int d = 0; // dont care

    static final int[][][] cornerLD_pattern =
        {    
        { { d, w, d },
          { e, w, w },
          { d, e, d },
          { d, d, d },
          { d, d, d } } };    
    
    static final int[][][] x_pattern =
        {
        { { d, w, d },
          { w, w, w },
          { e, w, d },
          { d, d, d },
          { d, d, d } },
        
        { { e, w, d },
          { w, w, w },
          { d, w, d },
          { d, d, d },
          { d, d, d } },
        
        { { d, w, e },
          { w, w, w },
          { d, w, d },
          { d, d, d },
          { d, d, d } },
        
        { { d, w, d },
          { w, w, w },
          { d, w, e },
          { d, d, d },
          { d, d, d } } };
    
    static final int[][][] cornerLU_pattern =
      { { { d, e, d },
          { e, w, w },
          { d, w, d },
          { d, d, d },
          { d, d, d } } };   
    
    static final int[][][] cornerRD_pattern =
      { { { d, w, d },
          { w, w, e },
          { d, e, d },
          { d, d, d },
          { d, d, d } } };   
    
    static final int[][][] cornerRU_pattern =
      { { { d, e, d },
          { w, w, e },
          { d, w, d },
          { d, d, d },
          { d, d, d } } };   
    
    static final int[][][] horizontal_pattern =
        {    
        { { d, e, d },
          { w, w, w },
          { d, e, d },
          { d, d, d },
          { d, d, d } } }; 
    
    static final int[][][] vertical_pattern =
        {   
        { { d, w, d },
          { e, w, e },
          { d, w, d },
          { d, d, d },
          { d, d, d } } }; 
    
    static final int[][][] hcapright_pattern =
        {   
        { { d, e, d },
          { w, w, e },
          { d, e, d },
          { d, d, d },
          { d, d, d } } }; 
    
    static final int[][][] hcapleft_pattern =
        {   
        { { d, e, d },
          { e, w, w },
          { d, e, d },
          { d, d, d },
          { d, d, d } } }; 
    
    static final int[][][] vcapup_pattern =
        {   
        { { d, e, d },
          { e, w, e },
          { d, w, d },
          { d, d, d },
          { d, d, d } } }; 
    
    static final int[][][] vcapdown_pattern =
        {   
        { { d, w, d },
          { e, w, e },
          { d, d, d },
          { d, d, d },
          { d, d, d } } }; 
    
    static final int[][][] tdown_pattern =
        {   
        { { d, w, d },
          { w, w, w },
          { d, e, d },
          { d, d, d },
          { d, d, d } } }; 

    static final int[][][] tup_pattern =
        {   
        { { d, e, d },
          { w, w, w },
          { d, w, d },
          { d, d, d },
          { d, d, d } } }; 
    
    static final int[][][] tleft_pattern =
        {   
        { { d, w, d },
          { e, w, w },
          { d, w, d },
          { d, d, d },
          { d, d, d } },
        
        { { d, w, d },
          { e, w, w },
          { d, w, d },
          { d, d, d },
          { d, d, d } }};     
    
    static final int[][][] tright_pattern =
        {   
        { { d, w, d },
          { w, w, e },
          { d, w, d },
          { d, d, d },
          { d, d, d } },
        
        { { d, w, d },
          { w, w, e },
          { d, w, d },
          { d, d, d },
          { d, d, d } } };   
    
    static final int[][][] cover_pattern =
      { { { d, d, d },
          { d, w, w },
          { d, w, w },
          { d, d, d },
          { d, d, d } } };
    
    static final int[][][] debug_pattern =
      { { { d, d, d },
          { d, w, d },
          { d, d, d },
          { d, d, d },
          { d, d, d } } };
    
    public enum Pattern {
        
        PCornerLD( DoodadType.CornerLD, cornerLD_pattern, true ),
        PCornerRD( DoodadType.CornerRD, cornerRD_pattern, true ),
        PCornerLU( DoodadType.CornerLU, cornerLU_pattern, true ),
        PCornerRU( DoodadType.CornerRU, cornerRU_pattern, true ),
        PHorizontal( DoodadType.Horizontal, horizontal_pattern, true ),
        PVertical( DoodadType.Vertical, vertical_pattern, true ),
        PCover( DoodadType.Cover, cover_pattern, false ),
        PCross( DoodadType.CrossWall, x_pattern, true ),
        PHCapRight( DoodadType.HCapRight, hcapright_pattern, true ),
        PHCapLeft( DoodadType.HCapLeft, hcapleft_pattern, true ),
        PVCapUp( DoodadType.VCapUp, vcapup_pattern, true ),
        PVCapDown( DoodadType.VCapDown, vcapdown_pattern, true ),
        PTDown( DoodadType.TDown, tdown_pattern, true ),
        PTUp( DoodadType.TUp, tup_pattern, true ),
        PTLeft( DoodadType.TLeft, tleft_pattern, true ),
        PTRight( DoodadType.TRight, tright_pattern, true );
        //PDebug( DoodadType.Spawn, debug_pattern, true );
        
        public DoodadType doodad;
        public int[][][] pattern;
        public boolean wall;
        Pattern( DoodadType doodad, int[][][] array, boolean wall ) {
            
            this.doodad = doodad;
            this.pattern = array;
            this.wall = wall;
        }
    }
    
    static public DoodadType searchPatterns( int x, int y, Tile[] tileArray, int width, boolean wall ) {
        
        for( Pattern p : Pattern.values() ) {
            
            if( p.wall == wall ) {
                
                for( int pi = 0; pi < p.pattern.length; pi++ ) {
                    
                    boolean match = true;
                    out:
                    for( int xi = 0; xi < 3; xi++ ) {
                        int xOffset = xi - 1;
                        for( int yi = 0; yi < 5; yi++ ) {
                           int yOffset = yi - 1; 

                           boolean isWall = true;
                           int index = x + xOffset + ( y + yOffset ) * width;
                           if( index < tileArray.length
                            && index >= 0 ) {
                               Tile tile = tileArray[ index ];
                               isWall = tile.wall;
                           }
                           
                            if( p.pattern[ pi ][ yi ][ xi ] != d )
                            {
                                 if( p.pattern[ pi ][ yi ][ xi ] == w ) {

                                     if( isWall ) {
                                         // match
                                     }
                                     else
                                     {
                                         // not a match
                                         match = false;
                                         break out;
                                     }
                                 } else {

                                     if( isWall ) {
                                         // not a match
                                         match = false;
                                         break out;
                                     }
                                     else
                                     {
                                         // match
                                     }
                                 }
                            }
                        }
                    }
                    if( match ) {
                        return p.doodad;
                    }
                }
            }
        }
        
        return null;
    }
        
}

File: src/generator/map/wallPattern.ts
import { Tile } from './tile'
import { DoodadTypeName } from '../objects/doodad'

// Pattern cell values (ported from WallPattern.java)
const w = 1 // wall
const e = 2 // empty
const d = 0 // don't care

type PatternGrid = number[][] // 5 rows x 3 cols
interface PatternDef {
  doodad: DoodadTypeName
  patterns: PatternGrid[]
  /** true = matches walls, false = matches floor decorations (cover) */
  wall: boolean
}

/**
 * 3x5 neighborhood patterns matched around every tile to pick which wall
 * doodad piece belongs there. Order matters: the first matching pattern wins,
 * exactly like the Java enum iteration order.
 */
const PATTERNS: PatternDef[] = [
  {
    doodad: 'CornerLD',
    wall: true,
    patterns: [
      [
        [d, w, d],
        [e, w, w],
        [d, e, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'CornerRD',
    wall: true,
    patterns: [
      [
        [d, w, d],
        [w, w, e],
        [d, e, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'CornerLU',
    wall: true,
    patterns: [
      [
        [d, e, d],
        [e, w, w],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'CornerRU',
    wall: true,
    patterns: [
      [
        [d, e, d],
        [w, w, e],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'Horizontal',
    wall: true,
    patterns: [
      [
        [d, e, d],
        [w, w, w],
        [d, e, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'Vertical',
    wall: true,
    patterns: [
      [
        [d, w, d],
        [e, w, e],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'Cover',
    wall: false,
    patterns: [
      [
        [d, d, d],
        [d, w, w],
        [d, w, w],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'CrossWall',
    wall: true,
    patterns: [
      [
        [d, w, d],
        [w, w, w],
        [e, w, d],
        [d, d, d],
        [d, d, d]
      ],
      [
        [e, w, d],
        [w, w, w],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ],
      [
        [d, w, e],
        [w, w, w],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ],
      [
        [d, w, d],
        [w, w, w],
        [d, w, e],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'HCapRight',
    wall: true,
    patterns: [
      [
        [d, e, d],
        [w, w, e],
        [d, e, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'HCapLeft',
    wall: true,
    patterns: [
      [
        [d, e, d],
        [e, w, w],
        [d, e, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'VCapUp',
    wall: true,
    patterns: [
      [
        [d, e, d],
        [e, w, e],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'VCapDown',
    wall: true,
    patterns: [
      [
        [d, w, d],
        [e, w, e],
        [d, d, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'TDown',
    wall: true,
    patterns: [
      [
        [d, w, d],
        [w, w, w],
        [d, e, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'TUp',
    wall: true,
    patterns: [
      [
        [d, e, d],
        [w, w, w],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'TLeft',
    wall: true,
    patterns: [
      [
        [d, w, d],
        [e, w, w],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  },
  {
    doodad: 'TRight',
    wall: true,
    patterns: [
      [
        [d, w, d],
        [w, w, e],
        [d, w, d],
        [d, d, d],
        [d, d, d]
      ]
    ]
  }
]

/**
 * Find the doodad for the tile at (x, y), or null if nothing matches.
 * Out-of-bounds neighbors count as walls, exactly like the original.
 */
export function searchPatterns(
  x: number,
  y: number,
  tileArray: Tile[],
  width: number,
  wall: boolean
): DoodadTypeName | null {
  for (const p of PATTERNS) {
    if (p.wall !== wall) continue

    for (const pattern of p.patterns) {
      let match = true

      outer: for (let xi = 0; xi < 3; xi++) {
        const xOffset = xi - 1
        for (let yi = 0; yi < 5; yi++) {
          const yOffset = yi - 1

          let isWall = true
          const index = x + xOffset + (y + yOffset) * width
          if (index >= 0 && index < tileArray.length) {
            isWall = tileArray[index].wall
          }

          const cell = pattern[yi][xi]
          if (cell === d) continue
          if (cell === w ? !isWall : isWall) {
            match = false
            break outer
          }
        }
      }

      if (match) {
        return p.doodad
      }
    }
  }

  return null
}
